File: iPhone/ActionSheet/ActionSheet.js
//
//  ActionSheet.js
//
// Created by Olivier Louvignes on 11/27/2011.
//
// Copyright 2011 Olivier Louvignes. All rights reserved.
// MIT Licensed

function ActionSheet() {}

ActionSheet.prototype.create = function(title, items, fn, options) {
	if(!options) options = {};

	var service = 'ActionSheet',
		action = 'create',
		callbackId = service + (PhoneGap.callbackId + 1);

	var config = {
		title : title+'' || '',
		items : items || ['Cancel'],
		callback : fn || function(){},
		scope: options.hasOwnProperty('scope') ? options.scope : null,
		style : options.hasOwnProperty('style') ? options.style+'' : 'default',
		destructiveButtonIndex : options.hasOwnProperty('destructiveButtonIndex') ? options.destructiveButtonIndex*1 : undefined,
		cancelButtonIndex : options.hasOwnProperty('cancelButtonIndex') ? options.cancelButtonIndex*1 : undefined
	};

	var callback = function(result) {
		var buttonValue = false, // value for cancelButton
			buttonIndex = result.buttonIndex;

		if(!config.cancelButtonIndex || buttonIndex != config.cancelButtonIndex) {
			buttonValue = config.items[buttonIndex];
		}

		config.callback.call(config.scope || null, buttonValue, buttonIndex);
	};

	PhoneGap.exec(callback, callback, service, action, [config]);
};

PhoneGap.addConstructor(function() {
	if(!window.plugins) window.plugins = {};
	window.plugins.actionSheet = new ActionSheet();
});

File: iOS/ActionSheet/ActionSheet.m
//
//  ActionSheet.m
//  
// Created by Olivier Louvignes on 11/27/2011.
// Added Cordova 1.5 support - @RandyMcMillan 2012
// Copyright 2011 Olivier Louvignes. All rights reserved.
// MIT Licensed

#import "ActionSheet.h" 

@implementation ActionSheet 

@synthesize callbackID;

-(void)create:(NSMutableArray*)arguments withDict:(NSMutableDictionary*)options  
{
	
	//NSLog(@"options: %@", options);
	//NSLog(@"arguments: %@", arguments);
	
	// The first argument in the arguments parameter is the callbackID.
	// We use this to send data back to the successCallback or failureCallback
	// through PluginResult.
	self.callbackID = [arguments pop];
	
	// Compiling options with defaults
	NSString *title = [options objectForKey:@"title"] ?: @"";
	NSString *style = [options objectForKey:@"style"] ?: @"black-translucent";
	//NSString *style = [options objectForKey:@"style"] ?: @"default";
	NSArray *items = [options objectForKey:@"items"];
	NSInteger cancelButtonIndex = [[options objectForKey:@"cancelButtonIndex"] intValue] ?: false;
	NSInteger destructiveButtonIndex = [[options objectForKey:@"destructiveButtonIndex"] intValue] ?: false;
	
	// Create actionSheet
	UIActionSheet *actionSheet = [[UIActionSheet alloc] initWithTitle:title
												   delegate:self
										  cancelButtonTitle:nil
									 destructiveButtonTitle:nil
										  otherButtonTitles:nil];
	
	// Style actionSheet, defaults to UIActionSheetStyleDefault
	if([style isEqualToString:@"black-opaque"]) actionSheet.actionSheetStyle = UIActionSheetStyleBlackOpaque;
	else if([style isEqualToString:@"black-translucent"]) actionSheet.actionSheetStyle = UIActionSheetStyleBlackTranslucent;
	else actionSheet.actionSheetStyle = UIActionSheetStyleDefault;
	
	// Fill with elements
	for(int i = 0; i < [items count]; i++) {
		[actionSheet addButtonWithTitle:[items objectAtIndex:i]];
	}
	// Handle cancelButtonIndex
	if([options objectForKey:@"cancelButtonIndex"]) {
		actionSheet.cancelButtonIndex = cancelButtonIndex;
	}
	// Handle destructiveButtonIndex
	if([options objectForKey:@"destructiveButtonIndex"]) {
		actionSheet.destructiveButtonIndex = destructiveButtonIndex;
	}

	// Toggle ActionSheet
    [actionSheet showInView:self.webView.superview];

}

/*-(void)show:(NSMutableArray*)arguments withDict:(NSMutableDictionary*)options  
{
	// Toggle ActionSheet
    [self.actionSheet showInView:self.webView.superview];
}*/

// ActionSheet generic dismiss
- (void)actionSheet:(UIActionSheet *)actionSheet didDismissWithButtonIndex:(NSInteger)buttonIndex
{
	//NSLog(@"didDismissWithButtonIndex:%d", buttonIndex);
	
	// Build returned result
	NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
	[result setObject:[NSNumber numberWithInteger:buttonIndex] forKey:@"buttonIndex"];
	
	// Create Plugin Result

	CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:result];
    
	// Checking if cancel was clicked
	if (buttonIndex != actionSheet.cancelButtonIndex) {
		//Call  the Failure Javascript function
		[self writeJavascript: [pluginResult toErrorCallbackString:self.callbackID]];
	// Checking if destructive was clicked
	} else if (buttonIndex != actionSheet.destructiveButtonIndex) {
		//Call  the Success Javascript function
		[self writeJavascript: [pluginResult toSuccessCallbackString:self.callbackID]];
	// Other button was clicked
	} else {    
		//Call  the Success Javascript function
		[self writeJavascript: [pluginResult toSuccessCallbackString:self.callbackID]];
	}
	
	// Release objects
	[actionSheet release];
}

@end
